Interaction: Use Rest Controller structure

obenland opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

Follow-up to #1146.

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* Updates Interactions endpoint to extend `WP_REST_Controller`.
* Adds unit tests for Interactions endpoint.

### Other information:

- [x] Have you written new tests for your changes, if applicable?

obenland committed:
Rename interaction endpoint file

obenland committed:
Update interaction controller endpoint

obenland committed:
Remove schema argument

This endpoint doesn't have a schema.

pfefferle commented on includes/rest/class-interaction-controller.php:
```diff
@@ -46,28 +51,19 @@ public static function register_routes() {
 		$uri          = $request->get_param( 'uri' );
 		$redirect_url = null;
-		$object       = Http::get_remote_object( $uri );
+		$object       = \Activitypub\Http::get_remote_object( $uri );
```
let's stick with using `use` for Namespaced classes. this is easier to read and consistent with the rest of the code.

obenland replied:
Fixed in fbbd9d5.

obenland committed:
Keep namespace import

Props @pfefferle

pfefferle committed:
Merge branch 'trunk' into update/interaction-endpoint

pfefferle committed:
Merge branch 'trunk' into update/interaction-endpoint

pfefferle commented on includes/rest/class-interaction-controller.php:
```diff
@@ -46,28 +53,19 @@ public static function register_routes() {
-			\is_wp_error( $object ) ||
-			! isset( $object['type'] )
-		) {
-			\wp_die(
```
this was a wp_die, because it might be user facing.

pfefferle replied:
what about adding a switch based on the accept header? if `text/html` return `wp_die` otherwise a `WP_Error`?

obenland replied:
I wondered about that! What's the scenario for when it's user-facing? How can I test that?

pfefferle replied:
go to a user on a mastodon instance, where you have no account on (for example https://mastodon.art/@pfefferle), click the follow button, then use your blogs WebFinger-ID in the follow dialogue.

pfefferle replied:
better build a custom URI instead of using an API endpoint?

pfefferle commented on includes/rest/class-interaction-controller.php:
```diff
@@ -104,31 +102,22 @@ public static function get( $request ) {
 
 		// Check if hook is implemented.
 		if ( ! $redirect_url ) {
-			\wp_die(
```
same as above

obenland committed:
Go back to using wp_die()

For requests that only accept json it'll return a json error response.

Props @pfefferle

obenland committed:
Add docs around this behavior.
